### PR TITLE
[func](jsonb)support json_depth function

### PR DESCRIPTION
**Pull request by LemonLiTree**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

File: docs/zh-CN/docs/sql-manual/sql-functions/json-functions/json-depth.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+`INT json_depth(JSON json_str)`
+
+返回 JSON 文档的最大深度。
+
+JSON_DEPTH()根据以下规则计算深度：
+
+* 空数组、空对象或标量的深度为 1。
+* 仅包含深度为 1 的元素的非空数组的深度为2。
+* 仅包含深度为 1 的成员值的非空对象的深度为2。
```
Comment (xiaokang) [inline docs/zh-CN/docs/sql-manual/sql-functions/json-functions/json-depth.md:39]: add explanation for depth > 2

File: gensrc/script/doris_builtins_functions.py [hunk 2]
```diff
@@ -1771,6 +1771,7 @@
         [['json_contains'], 'BOOLEAN', ['JSONB', 'JSONB'], 'ALWAYS_NULLABLE'],
         [['json_contains'], 'BOOLEAN', ['JSONB', 'JSONB', 'VARCHAR'], 'ALWAYS_NULLABLE'],
         [['json_contains'], 'BOOLEAN', ['JSONB', 'JSONB', 'STRING'], 'ALWAYS_NULLABLE'],
+        [['json_depth'], 'INT', ['JSONB'], 'ALWAYS_NULLABLE'],
```
Comment (xiaokang) [inline gensrc/script/doris_builtins_functions.py:1774]: why not depends on argument?

File: be/src/util/jsonb_document.h [hunk 3]
```diff
@@ -1264,6 +1267,56 @@ inline int JsonbValue::length() const {
...
+    case JsonbType::T_True:
+    case JsonbType::T_False: {
+        return 1;
+    }
+    case JsonbType::T_Object: {
+        int depth = 1;
+        int numElem = ((ObjectVal*)this)->numElem();
+        if (numElem == 0) return depth;
+
+        JsonbKeyValue* first_key = ((ObjectVal*)this)->getJsonbKeyValue(0);
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:1291]: why do special check for first element?

Comment (xiaokang) [reply be/src/util/jsonb_document.h:1291]: just set max_depth = depth and for loop from 0

File: be/src/util/jsonb_document.h [hunk 4]
```diff
@@ -1264,6 +1267,56 @@ inline int JsonbValue::length() const {
...
+    case JsonbType::T_Int128:
+    case JsonbType::T_String:
+    case JsonbType::T_Binary:
+    case JsonbType::T_Null:
+    case JsonbType::T_True:
+    case JsonbType::T_False: {
+        return 1;
+    }
+    case JsonbType::T_Object: {
+        int depth = 1;
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:1287]: suggest a better name base_depth

File: be/src/vec/functions/function_jsonb.cpp [hunk 5]
```diff
@@ -1288,6 +1288,56 @@ struct JsonbContainsAndPathImpl {
...
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) override {
+        DCHECK_GE(arguments.size(), 1);
+        auto jsonb_data_column = block.get_by_position(arguments[0]).column;
+
+        auto null_map = ColumnUInt8::create(input_rows_count, 0);
+        auto return_type = block.get_data_type(result);
+        MutableColumnPtr res = return_type->create_column();
+
+        //if(jsonb_data_column->)
```
Comment (xiaokang) [inline be/src/vec/functions/function_jsonb.cpp:1314]: delete unused code

File: be/src/vec/functions/function_jsonb.cpp [hunk 6]
```diff
@@ -1288,6 +1288,56 @@ struct JsonbContainsAndPathImpl {
...
+            // doc is NOT necessary to be deleted since JsonbDocument will not allocate memory
+            JsonbDocument* doc = JsonbDocument::createDocument(jsonb_value.data, jsonb_value.size);
+            JsonbValue* value = doc->getValue();
+            if (UNLIKELY(jsonb_value.size == 0 || !value)) {
+                null_map->get_data()[i] = 1;
+                res->insert_data(nullptr, 0);
+                continue;
+            }
+            auto depth = value->depth();
+            res->insert_data(const_cast<const char*>((char*)&depth), 0);
```
Comment (xiaokang) [inline be/src/vec/functions/function_jsonb.cpp:1333]: it's unsafe. Add a int32 to column is safe and efficient.

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

File: be/src/util/jsonb_document.h [hunk 7]
```diff
@@ -1269,6 +1272,53 @@ inline int JsonbValue::length() const {
...
+            int current_depth = depth + value->base_depth();
+            if (current_depth > max_depth) max_depth = current_depth;
+        }
+        return max_depth;
+    }
+    case JsonbType::T_Array: {
+        int depth = 1;
+        int numElem = ((ArrayVal*)this)->numElem();
+        if (numElem == 0) return depth;
+        int max_depth = depth + ((ArrayVal*)this)->get(0)->base_depth();
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:1310]: change as object, do not need special code for item 0

File: be/src/util/jsonb_document.h [hunk 8]
```diff
@@ -554,6 +554,9 @@ class JsonbValue {
     //Get the number of jsonbvalue elements
     int length() const;
 
+    //Get the depth of jsonbvalue
+    int base_depth() const;
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:558]: depth() is more simple and easy to understand

File: be/src/vec/functions/function_jsonb.cpp [hunk 9]
```diff
@@ -1287,6 +1287,57 @@ struct JsonbContainsAndPathImpl {
...
+            auto jsonb_value = jsonb_data_column->get_data_at(i);
+            // doc is NOT necessary to be deleted since JsonbDocument will not allocate memory
+            JsonbDocument* doc = JsonbDocument::createDocument(jsonb_value.data, jsonb_value.size);
+            JsonbValue* value = doc->getValue();
+            if (UNLIKELY(jsonb_value.size == 0 || !value)) {
+                null_map->get_data()[i] = 1;
+                res_data[i] = 0;
+                continue;
+            }
+            auto depth = value->base_depth();
```
Comment (xiaokang) [inline be/src/vec/functions/function_jsonb.cpp:1331]: unnessary variable depth

**Comment by LemonLiTree** [comment]
run buildall

File: docs/en/docs/sql-manual/sql-functions/json-functions/json-depth.md
Comment (morrySnow) [inline docs/en/docs/sql-manual/sql-functions/json-functions/json-depth.md:1]: doc not in this repo anymore. u need submit a new pull request in https://github.com/apache/doris-website

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonDepth.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * ScalarFunction 'json_depth'. This class is generated by GenerateFunction.
+ */
+public class JsonDepth extends ScalarFunction
+        implements BinaryExpression, ExplicitlyCastableSignature, AlwaysNullable {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonDepth.java:38]: JsonDepth is a UnaryExpression, not a BinaryExpression

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by zy-kkk — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonDepth.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * ScalarFunction 'json_depth'. This class is generated by GenerateFunction.
+ */
+public class JsonDepth extends ScalarFunction
+        implements BinaryExpression, ExplicitlyCastableSignature, PropagateNullable {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonDepth.java:38]: jsonDepth should be UnaryExpression

**Pull request closed without merging**